src/vm_pages.rs

```diff
@@ -732,29 +737,72 @@ pub struct VmPages<T: GuestStagePagingMode, S = VmStateFinalized> {
     pte_pages: PtePagePool,
     imsic_geometry: Once<GuestImsicGeometry>,
     iommu_context: Once<VmIommuContext>,
-    phantom: PhantomData<S>,
 }
 
-impl<T: GuestStagePagingMode, S> VmPages<T, S> {
+impl<T: GuestStagePagingMode> VmPages<T> {
+    /// Creates a new `VmPages` from the given root page table.
+    pub fn new(root: GuestStagePageTable<T>, regions: VmRegionList, nesting: usize) -> Self {
+        let page_tracker = root.page_tracker();
+        Self {
+            page_owner_id: root.page_owner_id(),
+            page_tracker: page_tracker.clone(),
+            tlb_tracker: TlbTracker::new(),
+            regions,
+            nesting,
+            root,
+            pte_pages: PtePagePool::new(page_tracker),
+            imsic_geometry: Once::new(),
+            iommu_context: Once::new(),
+        }
+    }
+
     /// Returns the `PageOwnerId` associated with the pages contained in this machine.
     pub fn page_owner_id(&self) -> PageOwnerId {
         self.page_owner_id
     }
 
-    /// Returns the address of the root page table for this VM.
-    pub fn root_address(&self) -> SupervisorPageAddr {
+    /// Returns a `VmPagesRef` to `self` in state `S`. The caller must ensure that `S` matches
```

Comment:
Do we have some way to enforce this safety requirement? Marking it as unsafe is probably overkill.

Reply:
Yeah, not sure it meets the definition of `unsafe` since it doesn't affect Salus' memory safety. Another way might be to hold a `&VmRef`, but that gets a little messy.

It's worth noting that getting the "wrong" type out of `as_ref()` would require explicitly naming the type, e.g. `let x = vm_pages.as_ref::<VmStateFinalized>()`, which should hopefully be an obvious code smell.